app/models/shipit/task.rb

```diff
@@ -47,6 +47,10 @@ def last_completed
         completed.last
       end
 
+      def previous_successful(id)
+        success.where('id < ?', id).last
```

Comment:
ActiveRecord question: is `last` doing this in the SQL, or are we returning the whole dataset here and discarding everything but the last element? Also can we trust implicit ordering here?

Reply:
last should add a `DESC LIMIT 1` to the query 🙂

Reply:
Any idea about the ordering Jack? Don't want this to accidentally return a really old deploy.

Reply:
I think this should work reliably, due to the nature of shipit creating new tasks on rollbacks/deploys. If anything, the success filter will give us a little more security as well. I don't anticipate any problems with this, and we've been using similar ID ordering in many other parts of shipit which have worked reliably

Reply:
`.last` gets baked into the SQL as a limit clause, for example, this _previous_successful_ call on a Task with id 197:
```sql
SELECT  "tasks".* FROM "tasks" WHERE "tasks"."type" IN ('Shipit::Deploy') AND "tasks"."status" = ? AND (id < 197) ORDER BY "tasks"."id" DESC LIMIT ?  [["status", "success"], ["LIMIT", 1]]
```
So it should be fine in that respect.

The ordering question is interesting. The default ordering by id should be fine (since we're already using) but most best practices blogs I've found recommend being explicit about order, and often, recommend using created_at and similar fields. However, in this context I don't see created_at being better or worse than the incrementing id.

I think the approach of being consistent and explicit makes sense, so adding an explicit ordering could be fine. but there are other places where we use .first and .last without explicitly ordering, so I'm not sure if that would be consistent.

What are your thoughts?

Reply:
If it's working reliably elsewhere then we don't need to go against the grain 👍